src/geometry/euclidean.lean

```diff
@@ -683,4 +683,230 @@ begin
+  exact (eq_of_vsub_eq_zero V
+          (submodule.disjoint_def.1
+            s.direction.orthogonal_disjoint
+            _
+            (vsub_mem_direction hp (orthogonal_projection_mem hn hc (v +ᵥ p)))
+            h)).symm
+end
+
+/-- Adding a vector to a point in the given subspace, then taking the
+orthogonal projection, produces the original point if the vector is a
+multiple of the result of subtracting a point's orthogonal projection
+from that point. -/
+lemma orthogonal_projection_vadd_smul_vsub_orthogonal_projection {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p1 : P} (p2 : P)
+    (r : ℝ) (hp : p1 ∈ s) :
+  orthogonal_projection hn hc (r • (p2 -ᵥ orthogonal_projection hn hc p2 : V) +ᵥ p1) = p1 :=
+orthogonal_projection_vadd_eq_self hn hc hp
+  (submodule.smul_mem _ _ (vsub_orthogonal_projection_mem_direction_orthogonal hn hc _))
+
+/-- The square of the distance from a point in `s` to `p` equals the
+sum of the squares of the distances of the two points to the
+`orthogonal_projection`. -/
+lemma dist_square_eq_dist_orthogonal_projection_square_add_dist_orthogonal_projection_square
+    {s : affine_subspace ℝ V P} (hn : (s : set P).nonempty)
+    (hc : is_complete (s.direction : set V)) {p1 : P} (p2 : P) (hp1 : p1 ∈ s) :
+  dist p1 p2 * dist p1 p2 =
+    dist p1 (orthogonal_projection hn hc p2) * dist p1 (orthogonal_projection hn hc p2) +
+    dist p2 (orthogonal_projection hn hc p2) * dist p2 (orthogonal_projection hn hc p2) :=
+begin
+  rw [metric_space.dist_comm p2 _, dist_eq_norm V p1 _, dist_eq_norm V p1 _, dist_eq_norm V _ p2,
+      ←vsub_add_vsub_cancel V p1 (orthogonal_projection hn hc p2) p2,
+      norm_add_square_eq_norm_square_add_norm_square_iff_inner_eq_zero],
+  exact submodule.inner_right_of_mem_orthogonal
+    (vsub_orthogonal_projection_mem_direction hn hc p2 hp1)
+    (orthogonal_projection_vsub_mem_direction_orthogonal hn hc p2)
+end
+
+/-- The square of the distance between two points constructed by
+adding multiples of the same orthogonal vector to points in the same
+subspace. -/
+lemma dist_square_smul_orthogonal_vadd_smul_orthogonal_vadd {s : affine_subspace ℝ V P}
+    {p1 p2 : P} (hp1 : p1 ∈ s) (hp2 : p2 ∈ s) (r1 r2 : ℝ) {v : V}
+    (hv : v ∈ s.direction.orthogonal) :
+  dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) * dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) =
+    dist p1 p2 * dist p1 p2 + (r1 - r2) * (r1 - r2) * (∥v∥ * ∥v∥) :=
+by rw [dist_eq_norm V (r1 • v +ᵥ p1), dist_eq_norm V p1, vsub_vadd_eq_vsub_sub, vadd_vsub_assoc,
+       add_comm, add_sub_assoc, ←sub_smul,
+       norm_add_square_eq_norm_square_add_norm_square
+         (submodule.inner_right_of_mem_orthogonal (vsub_mem_direction hp1 hp2)
+                                                  (submodule.smul_mem _ _ hv)),
+       norm_smul, real.norm_eq_abs, ←mul_assoc, mul_comm _ (abs (r1 - r2)), ←mul_assoc,
+       abs_mul_abs_self, ←mul_assoc]
```

Comment:
Could you make this into a `calc`-style proof? It's a bit hard to see what's going on now.

Reply:
Well, mathematically there's almost nothing going on, just some routine algebra with one use of Pythagoras. If we had tactics that could handle routine manipulations involving such things as norms, distances and `abs`, it would be natural to use those to make it shorter. I'm not sure making it longer by showing each (uninteresting) intermediate expression explicitly in a `calc` proof actually makes it any clearer compared to leaving it to people to look at the tactic state after each comma if they wish to see the intermediate expressions.

Reply:
There's definitely no readability to be gained by writing out each `rw` step on its own line. My thoughts were more along the lines of:

```lean
/-- The square of the distance between two points constructed by
adding multiples of the same orthogonal vector to points in the same
subspace. -/
lemma dist_square_smul_orthogonal_vadd_smul_orthogonal_vadd {s : affine_subspace ℝ V P}
    {p1 p2 : P} (hp1 : p1 ∈ s) (hp2 : p2 ∈ s) (r1 r2 : ℝ) {v : V}
    (hv : v ∈ s.direction.orthogonal) :
  dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) * dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) =
    dist p1 p2 * dist p1 p2 + (r1 - r2) * (r1 - r2) * (∥v∥ * ∥v∥) :=
calc dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) * dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2)
    = ∥(p1 -ᵥ p2) + (r1 - r2) • v∥ * ∥(p1 -ᵥ p2) + (r1 - r2) • v∥
  : by { rw [dist_eq_norm V (r1 • v +ᵥ p1), vsub_vadd_eq_vsub_sub, vadd_vsub_assoc, sub_smul], abel }
... = ∥p1 -ᵥ p2∥ * ∥p1 -ᵥ p2∥ + ∥(r1 - r2) • v∥ * ∥(r1 - r2) • v∥
  : norm_add_square_eq_norm_square_add_norm_square
      (submodule.inner_right_of_mem_orthogonal (vsub_mem_direction hp1 hp2)
        (submodule.smul_mem _ _ hv))
... = ∥(p1 -ᵥ p2 : V)∥ * ∥(p1 -ᵥ p2 : V)∥ + abs (r1 - r2) * abs (r1 - r2) * ∥v∥ * ∥v∥
  : by { rw [norm_smul, real.norm_eq_abs], ring }
... = dist p1 p2 * dist p1 p2 + (r1 - r2) * (r1 - r2) * (∥v∥ * ∥v∥)
  : by { rw [dist_eq_norm V p1, abs_mul_abs_self], ring }
```
I feel the calc-style proof makes it more clear that the main argument is `norm_add_square_eq_norm_square_add_norm_square`, along with some manipulation of `dist`, `norm` and `abs`, along with some simple algebra. Unfortunately, we need to help the `abel` and `ring` tactics a lot if we want to do this, so I'll leave it up to you if it really is an improvement.

Reply:
Used that proof.